frontend/src/pages/users/+Page.tsx

```diff
@@ -0,0 +1,232 @@
+        console.error('Fetch error', error)
+      })
+    }
+  }, [authContext.isAdmin])
+
+  const fetchUsers = async (): Promise<void> => {
+    const response = await fetch('/api/user')
+    if (!response.ok) {
+      throw new Error('Failed to get users')
+    }
+    const data: User[] = await response.json()
+    console.log(data)
+    setUsers(data)
+  }
+
+  const addUser = async (): Promise<void> => {
+    const { email, isAdmin, isManager } = newUser
+
+    if (email.match(/^\S+@rice\.edu$/) == null) {
+      setError('Please enter a valid Rice University email')
+      return
+    }
+    try {
+      const response = await fetch('/api/user', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({ email, is_admin: isAdmin, is_manager: isManager })
+      })
+
+      if (!response.ok) {
+        setError('Failed to add user')
+        throw new Error('Failed to add user')
+      }
+
+      const data: User = await response.json()
+      setUsers([...users, data])
+      setNewUser({ email: '', isAdmin: false, isManager: false })
+      setError('')
+    } catch (error) {
+      console.error('Add user error', error)
+    }
+  }
+
+  const deleteUser = async (id: number): Promise<void> => {
+    try {
+      const response = await fetch(`/api/user/${id}`, {
+        method: 'DELETE'
+      })
+
+      if (!response.ok) {
+        setError('Failed to delete user')
+        throw new Error('Failed to delete user')
+      }
+
+      setUsers(users.filter(user => user.id !== id))
+    } catch (error) {
+      console.error('Delete user error', error)
+    }
+  }
+
+  const updateUserPermissions = async (id: number, isAdmin: boolean, isManager: boolean, email: string): Promise<void> => {
+    try {
+      const response = await fetch(`/api/user/${id}`, {
+        method: 'PUT',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({ id, is_admin: isAdmin, is_manager: isManager, email })
+      })
+
+      if (!response.ok) {
+        setError('Failed to update user permissions')
+        throw new Error('Failed to update user permissions')
+      }
+
+      setUsers(users.map(user => {
+        if (user.id === id) {
+          return { ...user, is_admin: isAdmin, is_manager: isManager, email }
+        }
+        return user
+      }))
+    } catch (error) {
+      console.error('Update user permissions error', error)
+    }
+  }
+
+  const renderUsers = users.length > 0
+    ? (
+        users.map(user => (
+        <tr key={user.email}>
+          <Td>{user.email}</Td>
+          <Td>
+            <input
+              type="checkbox"
+              name="isAdmin"
+              checked={user.is_admin}
+              onChange={(e) => {
+                const isChecked = e.target.checked
+                void updateUserPermissions(user.id, isChecked, user.is_manager, user.email)
```

Comment:
Minor: I don't personally care too much since this is just going to impact devs, but it wasn't immediately obvious to me when toggling these that something was happening since there wasn't any UI indication. I did finally figure it out :joy: .

Reply:
Same thought - I didn't read any of Ripal's comments before reviewing and had the same thought. I figured out I was in fact updating the roles by using the check boxes.